### PR TITLE
Use WP_Query in get_page_by_path.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket:  https://core.trac.wordpress.org/ticket/56689
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/post.php [hunk 1]
```diff
@@ -5674,52 +5672,41 @@ function get_page( $page, $output = OBJECT, $filter = 'raw' ) {
...
-	$sql                 = "
-		SELECT ID, post_name, post_parent, post_type
-		FROM $wpdb->posts
-		WHERE post_name IN ($in_string)
-		AND post_type IN ($post_type_in_string)
-	";
+	$args = array(
+		'post_name__in'          => $parts,
+		'post_type'              => $post_types,
+		'post_status'            => 'all',
```
Comment (spacedmonkey) [inline src/wp-includes/post.php:5686]: @peterwilsoncc Fun fact this work!
Screenshot 2022-12-13 at 12 53 17

Comment (peterwilsoncc) [reply src/wp-includes/post.php:5686]: Possibly a manifestation of https://core.trac.wordpress.org/ticket/51094

File: src/wp-includes/post.php [hunk 2]
```diff
@@ -5670,52 +5668,41 @@ function get_page( $page, $output = OBJECT, $filter = 'raw' ) {
...
 	}
 
-	$post_types          = esc_sql( $post_types );
-	$post_type_in_string = "'" . implode( "','", $post_types ) . "'";
-	$sql                 = "
-		SELECT ID, post_name, post_parent, post_type
-		FROM $wpdb->posts
-		WHERE post_name IN ($in_string)
-		AND post_type IN ($post_type_in_string)
-	";
```
Comment (peterwilsoncc) [inline src/wp-includes/post.php:5711]: I just noticed this is another one of those jolly queries without an orderby.

File: src/wp-includes/post.php (hunk 1 above)
Comment (spacedmonkey) [reply src/wp-includes/post.php:5686]: @peterwilsoncc How about adding a new special case for post status, where if you are give `all`, it returns all. This maybe confusing with `any`. But it clear to me that we need a state to get all posts of any post status, there is at least two places in core that need it.

File: src/wp-includes/post.php (hunk 2 above)
Comment (spacedmonkey) [reply src/wp-includes/post.php:5711]: I do `'orderby'                => 'none',`.

File: src/wp-includes/post.php (hunk 1 above)
Comment (spacedmonkey) [reply src/wp-includes/post.php:5686]: We could also do the same workaround as in https://github.com/WordPress/wordpress-develop/commit/7ad9e7d04e70c1f63c024ef58065aa49ca30eab0.

File: tests/phpunit/tests/post/getPageByPath.php [hunk 3]
```diff
@@ -103,6 +103,85 @@ public function test_should_match_nested_page() {
...
+		);
+
+		$queries_before = get_num_queries();
+		$found          = get_page_by_path( 'foo/bar/baz' );
+		$queries_after  = get_num_queries();
+		$cached_post    = wp_cache_get( $p1, 'posts' );
+
+		$this->assertSame( 1, $queries_after - $queries_before, 'Only one query should run' );
+		$this->assertSame( $p3, $found->ID, 'Check to see if the result is correct' );
+		$this->assertIsObject( $cached_post, 'The cached post is not an object' );
```
Comment (costdev) [inline tests/phpunit/tests/post/getPageByPath.php:142]: Should `$cached_post` be an instance of `WP_Post`, or could it be any object?

If `WP_Post`, we can use `$this->assertInstanceOf( 'WP_Post', $cached_post, 'The cached post is not a WP_Post object' );`

Same question for the second test too

**Review by costdev — COMMENTED** [review]
Thanks for adding the tests @spacedmonkey! I just have one question about the expected type of `$cached_post` in the tests. :slightly_smiling_face:

File: tests/phpunit/tests/post/getPageByPath.php (hunk 3 above)
Comment (spacedmonkey) [inline tests/phpunit/tests/post/getPageByPath.php:142]:
```suggestion
		$this->assertInstanceOf( 'WP_Post', $cached_post, 'The cached post is not a WP_Post object' );
```

File: tests/phpunit/tests/post/getPageByPath.php [hunk 4]
```diff
@@ -103,6 +103,85 @@ public function test_should_match_nested_page() {
...
```
Comment (spacedmonkey) [inline tests/phpunit/tests/post/getPageByPath.php:182]:
```suggestion
		$this->assertInstanceOf( 'WP_Post', $cached_post, 'The cached post is not a WP_Post object' );
```

**Review by costdev — APPROVED** [review]
Thanks for the update @spacedmonkey! LGTM :+1:

File: tests/phpunit/tests/post/getPageByPath.php [hunk 5]
```diff
@@ -103,6 +103,85 @@ public function test_should_match_nested_page() {
...
+		);
+
+		$queries_before = get_num_queries();
+		$found          = get_page_by_path( 'foo/bar/baz' );
+		$queries_after  = get_num_queries();
+		$cached_post    = wp_cache_get( $p1, 'posts' );
+
+		$this->assertSame( 1, $queries_after - $queries_before, 'Only one query should run' );
+		$this->assertSame( $p3, $found->ID, 'Check to see if the result is correct' );
+		$this->assertInstanceOf( 'WP_Post', $cached_post, 'The cached post is not a WP_Post object' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/getPageByPath.php]:
```suggestion
		$this->assertIsObject( $cached_post, 'The cached post is not an object' );
```

File: tests/phpunit/tests/post/getPageByPath.php [hunk 6]
```diff
@@ -103,6 +103,85 @@ public function test_should_match_nested_page() {
...
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/getPageByPath.php]:
```suggestion
		$this->assertIsObject( $cached_post, 'The cached post is not an object' );
```

File: tests/phpunit/tests/post/getPageByPath.php [hunk 7]
```diff
@@ -103,6 +103,85 @@ public function test_should_match_nested_page() {
...
+	}
+
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/getPageByPath.php:200]: Some additional tests for unhappy paths are needed too. 

This is a starting point, it passes on trunk.

```suggestion

	/**
	 * @ticket 56689
	 *
	 * @covers ::get_page_by_path
	 */
	public function test_should_return_null_for_invalid_path() {
		$queries_before = get_num_queries();
		$get_1          = get_page_by_path( 'should/return/null/for/an/invalid/path' );
		$get_2          = get_page_by_path( 'should/return/null/for/an/invalid/path' );
		$queries_after  = get_num_queries();

		$this->assertNull( $get_1, 'Invalid path should return null.' );
		$this->assertSame( 1, $queries_after - $queries_before, 'Only one query should run.' );
		$this->assertSame( $get_1, $get_2, 'The cached result should be the same as the uncached result.' );
	}

```

**Review by peterwilsoncc — COMMENTED** [review]
The source changes LGTM.

The earlier assertion was correct, posts are cached as standard objects and converted to `WP_Post` objects within `WP_Post::get_instance()`.

https://github.com/WordPress/wordpress-develop/blob/5ebe28966e5decbe1862c147bf98db0d8094038e/src/wp-includes/class-wp-post.php#L255

That's the cause of the failing tests. The other tests pass on trunk.

I've also suggested some unhappy (or less happy) path tests. I couldn't see anything testing for `null` when expected, but I am sure there are other situations that can be included.

**Comment by spacedmonkey** [comment]
Committed https://github.com/WordPress/wordpress-develop/commit/4ca85e7528a42f2a6e3b5f6c942e5486c471ca78

**Pull request closed without merging**